File: tests/fixtures/sample.ts
/**
 * Sample TypeScript file for testing
 * Contains functions, classes, imports, exports
 */

import { Request, Response } from 'express';
import * as path from 'node:path';
import { DatabaseClient } from './database';
import type { User, UserRole } from '@/types/user';

// Top-level constant
export const API_VERSION = 'v1';
export const MAX_RETRIES = 3;

/**
 * User authentication service
 * Handles login, logout, and session management
 */
export class AuthService {
  private dbClient: DatabaseClient;
  private sessionTimeout: number;

  constructor(dbClient: DatabaseClient) {
    this.dbClient = dbClient;
    this.sessionTimeout = 3600000; // 1 hour
  }

  /**
   * Authenticate user with credentials
   *
   * @param email - User email address
   * @param password - User password
   * @returns User object if authentication succeeds
   */
  async login(email: string, password: string): Promise<User | null> {
    if (!email || !password) {
      throw new Error('Email and password are required');
    }

    const user = await this.dbClient.query(
      'SELECT * FROM users WHERE email = $1',
      [email]
    );

    if (!user) {
      return null;
    }

    const isValid = await this.verifyPassword(password, user.password_hash);

    if (!isValid) {
      return null;
    }

    return user;
  }

  /**
   * Verify password against hash
   */
  private async verifyPassword(password: string, hash: string): Promise<boolean> {
    // Simulate password verification logic
    return password === hash; // Simplified for testing
  }

  /**
   * Create new user session
   */
  async createSession(userId: string): Promise<string> {
    const sessionId = this.generateSessionId();

    await this.dbClient.query(
      'INSERT INTO sessions (id, user_id, expires_at) VALUES ($1, $2, $3)',
      [sessionId, userId, new Date(Date.now() + this.sessionTimeout)]
    );

    return sessionId;
  }

  /**
   * Generate random session ID
   */
  private generateSessionId(): string {
    return Math.random().toString(36).substring(2, 15);
  }
}

/**
 * Express route handler for login
 */
export const loginHandler = async (req: Request, res: Response): Promise<void> => {
  const { email, password } = req.body;

  try {
    const authService = new AuthService(req.app.get('dbClient'));
    const user = await authService.login(email, password);

    if (!user) {
      res.status(401).json({ error: 'Invalid credentials' });
      return;
    }

    const sessionId = await authService.createSession(user.id);

    res.cookie('session_id', sessionId, {
      httpOnly: true,
      secure: process.env.NODE_ENV === 'production',
      maxAge: 3600000,
    });

    res.json({ user });
  } catch (error) {
    console.error('Login error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
};

/**
 * Check user role and permissions
 */
export function hasPermission(user: User, requiredRole: UserRole): boolean {
  const roleHierarchy: Record<UserRole, number> = {
    admin: 3,
    moderator: 2,
    user: 1,
  };

  return roleHierarchy[user.role] >= roleHierarchy[requiredRole];
}

// Default export
export default AuthService;

File: tests/fixtures/sample.py
"""
Sample Python file for testing
Contains functions, classes, imports
"""

import os
import sys
from typing import Optional, List, Dict
from datetime import datetime, timedelta
import numpy as np
from utils.database import Database
from models.user import User, UserRole

# Module-level constants
API_VERSION = "v1"
MAX_RETRIES = 3


class AuthService:
    """
    User authentication service
    Handles login, logout, and session management
    """

    def __init__(self, db_client: Database):
        """Initialize auth service with database client"""
        self.db_client = db_client
        self.session_timeout = 3600  # 1 hour in seconds

    async def login(self, email: str, password: str) -> Optional[User]:
        """
        Authenticate user with credentials

        Args:
            email: User email address
            password: User password

        Returns:
            User object if authentication succeeds, None otherwise
        """
        if not email or not password:
            raise ValueError("Email and password are required")

        user = await self.db_client.query(
            "SELECT * FROM users WHERE email = ?", (email,)
        )

        if not user:
            return None

        is_valid = await self._verify_password(password, user.password_hash)

        if not is_valid:
            return None

        return user

    async def _verify_password(self, password: str, hash: str) -> bool:
        """Verify password against hash"""
        # Simplified for testing
        return password == hash

    async def create_session(self, user_id: str) -> str:
        """Create new user session"""
        session_id = self._generate_session_id()
        expires_at = datetime.now() + timedelta(seconds=self.session_timeout)

        await self.db_client.query(
            "INSERT INTO sessions (id, user_id, expires_at) VALUES (?, ?, ?)",
            (session_id, user_id, expires_at),
        )

        return session_id

    def _generate_session_id(self) -> str:
        """Generate random session ID"""
        import random
        import string

        return "".join(random.choices(string.ascii_letters + string.digits, k=32))


def has_permission(user: User, required_role: UserRole) -> bool:
    """
    Check user role and permissions

    Args:
        user: User object
        required_role: Required role for access

    Returns:
        True if user has required permission
    """
    role_hierarchy = {
        UserRole.ADMIN: 3,
        UserRole.MODERATOR: 2,
        UserRole.USER: 1,
    }

    return role_hierarchy.get(user.role, 0) >= role_hierarchy.get(required_role, 0)


def calculate_complexity(data: List[int]) -> Dict[str, float]:
    """
    Calculate statistical complexity metrics

    Demonstrates loop and conditional usage
    """
    if not data:
        return {"mean": 0.0, "std": 0.0, "max": 0.0}

    total = 0
    count = len(data)

    # Calculate mean
    for value in data:
        total += value

    mean = total / count

    # Calculate standard deviation
    variance_sum = 0
    for value in data:
        variance_sum += (value - mean) ** 2

    std = (variance_sum / count) ** 0.5

    # Find maximum
    max_value = data[0]
    for value in data:
        if value > max_value:
            max_value = value

    return {"mean": mean, "std": std, "max": float(max_value)}


# Exports
__all__ = ["AuthService", "has_permission", "calculate_complexity"]
